File: motio-commons/src/main/java/com/motio/commons/model/NotificationMessage.java
package com.motio.commons.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.ZonedDateTime;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "notification_message")
public class NotificationMessage {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column
    private String firebaseGeneratedId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private MessageType messageType;

    @ManyToOne
    @JoinColumn(name = "receiver_id")
    private User receiver;  // null == broadcast message

    @Column
    private String title;

    @Column
    private String body;

    @Column
    private ZonedDateTime sendDateTime;
}

File: motio_client/lib/models/notification_message.dart
import 'package:motio_client/models/user.dart';

class NotificationMessage {
  final int id;
  final String? firebaseGeneratedId;
  final String messageType;
  final User? receiver;
  final String title;
  final String body;
  final DateTime sendDateTime;

  NotificationMessage({
    required this.id,
    this.firebaseGeneratedId,
    required this.messageType,
    this.receiver,
    required this.title,
    required this.body,
    required this.sendDateTime,
  });

  factory NotificationMessage.fromJson(Map<String, dynamic> json) {
    return NotificationMessage(
      id: json['id'],
      firebaseGeneratedId: json['firebaseGeneratedId'],
      messageType: json['messageType'],
      receiver: User.fromJson(json['receiver']),
      title: json['title'],
      body: json['body'],
      sendDateTime: DateTime.parse(json['sendDateTime']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'firebaseGeneratedId': firebaseGeneratedId,
      'messageType': messageType,
      'receiver': receiver,
      'title': title,
      'body': body,
      'sendDateTime': sendDateTime.toIso8601String(),
    };
  }
}
